consensus/src/lib.rs

```diff
@@ -155,8 +174,48 @@ impl Consensus {
                     (preferred_id, preferred_seq)
                 );
 
+                //adjust batch pool
+                //put back batches from latest proposal if have started deliberation
+                match self.proposals.entry(self.latest_ledger.id) {
+                    Entry::Occupied(mut proposals) => {
+                        match proposals.get_mut().entry(self.node_id) {
+                            Entry::Occupied(proposal) => {
+                                let to_q = proposal.get().proposal.batches.clone();
+                                info!("adjust pool adding {:?} proposal batches", to_q.len());
+                                self.batch_pool.extend(to_q);
+                            }
+                            Entry::Vacant(_) => {}
+                        }
+                    }
+                    Entry::Vacant(_) => {}
+                };
+
+                //adjust pool by adding batches on the wrong branch and removing from right branch
+                //by current design (oct 31, 2023), the node can be off by one ledger, not more
+
+                //put back batches in the ledgers on the wrong branch
+                let to_q = self.latest_ledger.batch_set.clone();
+                info!("adjust pool adding {:?} ledger batches", to_q.len());
+                self.batch_pool.extend(to_q);
+
+                //latest's parent is the common ancestor of the branches
+                let common_ancestor = self.latest_ledger.ancestors[0].clone();
+                //find out batches in the ledgers on the right branch, and take them out
+                let mut l = self.validations.adaptor_mut().acquire(&preferred_id).await
+                    .expect("ValidationsAdaptor did not have a ledger in cache.");
+                while l.id != common_ancestor {
+                    l.batch_set.iter().for_each(|x| {self.negative_pool.insert(x.clone());});
+                    l = self.validations.adaptor_mut().acquire(&l.ancestors[0]).await
+                    .expect("ValidationsAdaptor did not have a ledger in cache.");
+                }
+                self.batch_pool = self.batch_pool.clone().into_iter().filter(|b| { !self.negative_pool.contains(b) }).collect();
```

Comment:
```suggestion
                self.batch_pool = self.batch_pool.iter()
                    .filter(|b| { !self.negative_pool.contains(b) })
                    .map(|b| *b)
                    .collect();
```

I think this is slightly better than cloning the entire batch_pool and then filtering. We're still making copies of the batches that we ultimately want, but at least we won't be copying the batches that get filtered out.